src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -1846,36 +1846,64 @@ private static void CheckStringSplitOptions(StringSplitOptions options)
 
         // Returns a substring of this string.
         //
-        public string Substring(int startIndex) => Substring(startIndex, Length - startIndex);
-
-        public string Substring(int startIndex, int length)
+        public string Substring(int startIndex)
         {
-            if (startIndex < 0)
+            if (startIndex == 0)
             {
-                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndex);
+                return this;
             }
 
-            if (startIndex > Length)
-            {
-                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_StartIndexLargerThanLength);
-            }
+            int length = Length;
+            // Copy Span.Slice checks for perf
+            if ((uint)startIndex > (uint)length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
```

Comment:
For compatibility, we'll need to throw the same exception (including the same argument name) as we used to. But I think you're on the right track here: handle known-good cases first using the minimum number of checks. Then if these checks fail, then go do the more expensive logic of figuring out exactly what exception should be thrown.

Remember also to special-case `startIndex == length` and to return `string.Empty`.